### ISSUE TITLE
Dots at the end of many messages are omitted

### ISSUE THREAD

**Issue by markelog** [eslint/eslint#6762]
**What version of ESLint are you using?**
3.1.1
**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
Example for `array-bracket-spacing` but this issue is not limited to this rule only - 

```json
{
	"rules": {
		"array-bracket-spacing": "error"
	}
}
```

**What did you do? Please include the actual source code causing the issue.**
```js
var foo = [ ]
```

**What did you expect to happen?**
Have a dot at the end - 
`  1:11  error  There should be no space after '['.  array-bracket-spacing`

**What actually happened? Please include the actual, raw output from ESLint.**
It doesn't have a dot at the end - 
`  1:11  error  There should be no space after '['  array-bracket-spacing`

--

I'm not sure if this is intentional or not, every message should have a dot at the end or there is a different policy in place? If so, I think it should be documented.

Beside that, if this is really a "bug" I think every rule should be fixed with one commit in order not to trash the commit history

**Comment by platinumazure** [comment]
As I noted in another issue, be careful to check this with a formatter that isn't "stylish", which does strip periods from all messages. (The ESLint demo is a sufficient tool for checking the error messages, as well as most if not all of the other formatters.)

**Comment by markelog** [comment]
What do you mean? Can't we just check the source? If we want to act on it, then I would prefer to add `beginner` label, this issue seems perfect for the first timer

**Comment by markelog** [comment]
Looking it from the other perspective we can just remove the dots from all the rules and re-add them with reporter, this would be an opposite to the `stylish` logic.

That way, we wouldn't need to "remember" check those dots when reviewing new PRs

**Comment by markelog** [comment]
> That way, we wouldn't need to "remember" check those dots when reviewing new PRs

With this path though, we would need to check for absence of those dots :), but I guess reporters could check for their presence, but that wouldn't work for custom reporters then, doesn't sound like a perfect solution

**Comment by pmcelhaney** [comment]
How about updating the reporter to add the dot if and only if it's missing? That way 3rd party rules would also be normalized. (Not sure if that's a big or a feature. :))

**Comment by markelog** [comment]
@pmcelhaney it still be inconsistent in the sources though.

btw, @wavebeem do you want to get in on this?

**Comment by pmcelhaney** [comment]
Yeah. 

catching manually in PR < fixing at runtime < catching at build time < catching and fixing with a custom ESLint rule 

Just not sure what's possible at this point.

**Comment by pmcelhaney** [comment]
Okay, maybe a check for proper formatting can be added to rule-tester.jd. 

https://github.com/eslint/eslint/blob/master/lib/testers/rule-tester.js#L407

Is there a way to add that check only to this repo but not affect third-party rules that doesn't affect the spirit of rules being independent?

**Comment by wavebeem** [comment]
Hmm, normalizing the presence/absence of `.` at the end of messages in code sounds pretty smart to me, especially considering custom rules are a core feature of ESLint. I dunno.

**Comment by vitorbal** [comment]
We could add a custom internal rule for checking that the last dot is
always present, similar to what we did for enforcing the correct structure
of the "meta" property in rule definitions.
I just wonder if that wouldn't be overkill.

**Comment by pmcelhaney** [comment]
I messed around with adding another `assert` at line 395 of rule-tester.js.

```js
                    if (messages[i].message) {
                        assert.equal(messages[i].message.slice(-1), ".", messages[i].message);
                    }
```

And found the following messages. If there's an interest, I may go through and normalize all of the messages in the source code this weekend. 

```txt
'A' is already defined
'A' was used before it was defined
'Object' is already defined
'a' is already defined
'a' was used before it was defined
'b' is already defined
'top' is already defined
A space is required after '['
A space is required after '{'
A space is required before ']'
A space is required before '}'
Avoid using Function.prototype.apply, instead use Reflect.apply
Avoid using Function.prototype.call, instead use Reflect.apply
Avoid using Object.defineProperty, instead use Reflect.defineProperty
Avoid using Object.getOwnPropertyDescriptor, instead use Reflect.getOwnPropertyDescriptor
Avoid using Object.getOwnPropertyNames, instead use Reflect.getOwnPropertyNames
Avoid using Object.getPrototypeOf, instead use Reflect.getPrototypeOf
Avoid using Object.isExtensible, instead use Reflect.isExtensible
Avoid using Object.preventExtensions, instead use Reflect.preventExtensions
Avoid using Object.setPrototypeOf, instead use Reflect.setPrototypeOf
Avoid using the delete keyword, instead use Reflect.deleteProperty
Do not nest ternary expressions
Don't make functions within a loop
Expected space or tab before '*/' in comment
Expected whitespace after rest operator
Expected whitespace after rest property operator
Expected whitespace after spread operator
Expected whitespace after spread property operator
File must be at most 2 lines long
Getter is not present
Identifier 'arr' is blacklisted
Identifier 'bar' is blacklisted
Identifier 'baz' is blacklisted
Identifier 'foo' is blacklisted
Identifier 'key' is blacklisted
Identifier 'myArray' is blacklisted
Identifier 'myDate' is blacklisted
Identifier 'obj' is blacklisted
Identifier name '_$x$_t$' is too long. (> 4)
Identifier name '_$xt_$' is too long. (> 4)
Identifier name 'a' is too short. (< 2)
Identifier name 'e' is too short. (< 2)
Identifier name 'i' is too short. (< 2)
Identifier name 'j' is too short. (< 2)
Identifier name 'x' is too short. (< 2)
Invalid flags supplied to RegExp constructor 'u'
Invalid flags supplied to RegExp constructor 'y'
Invalid flags supplied to RegExp constructor 'yu'
Invalid flags supplied to RegExp constructor 'z'
Invalid regular expression: /)/: Unmatched ')'
Invalid regular expression: /[/: Unterminated character class
Invalid typeof comparison value
Irregular whitespace not allowed
Missing '()' invoking a constructor
Missing space before =>
No magic number: -60
No magic number: 0
No magic number: 0x1A
No magic number: 1
No magic number: 10
No magic number: 2
No magic number: 3
No magic number: 4
No magic number: 42
No magic number: 5
No magic number: 60
Number constants declarations must use 'const'
Object properties must go on a new line
Object properties must go on a new line if they aren't all on the same line
Setter is not present
The 'in' expression's left operand is negated
There should be no line break before or after '+'
There should be no line break before or after '+='
There should be no line break before or after '='
There should be no line break before or after '?'
There should be no line break before or after '||'
There should be no space after '['
There should be no space after '{'
There should be no space before ']'
There should be no space before '}'
Unexpected parentheses around single function argument
Unexpected space after =>
Unexpected space before =>
Unexpected space or tab before '*/' in comment
Unexpected use of 'event'
Unexpected use of 'foo'
Unexpected use of continue statement
Unexpected whitespace after rest operator
Unexpected whitespace after rest property operator
Unexpected whitespace after spread operator
Unexpected whitespace after spread property operator
Unnecessary escape character: \"
Unnecessary escape character: \#
Unnecessary escape character: \$
Unnecessary escape character: \'
Unnecessary escape character: \;
Unnecessary escape character: \@
Unnecessary escape character: \B
Unnecessary escape character: \a
Unnecessary escape character: \d
Unnecessary escape character: \p
Unnecessary use of boolean literals in conditional expression
Unnecessary use of conditional expression for default assignment
Unsafe usage of BreakStatement
Unsafe usage of ContinueStatement
Unsafe usage of ReturnStatement
Unsafe usage of ThrowStatement
Variables within the same declaration block should be sorted alphabetically
expected [Function] to not throw an error but 'Invalid args' was thrown
expected [Function] to throw error including 'Rule should not modify AST.' but got 'error'
```

While I'm at it I'll go ahead and capitalize "expected". :)

**Comment by markelog** [comment]
@vitorbal 

> We could add a custom internal rule for checking that the last dot is
always present

You mean custom ESLint rule? I'm not sure if that is possible for static analysis tool, for example, you would need cases like that - https://github.com/eslint/eslint/blob/ddea63a6dbc73e9ef2c2c5625826d02d586de774/lib/rules/array-bracket-spacing.js#L78

@pmcelhaney

> Okay, maybe a check for proper formatting can be added to rule-tester.js.

I think this is our best bet, it shouldn't reflect on rules which are not part of our repo, so we wouldn't need to bump major or anything like that.

**Comment by pmcelhaney** [comment]
@markelog The change to rule-tester.js is simple, but how do we prevent it from affecting third party rules? Don't they use rule-tester.js in their tests? 

I was thinking maybe add a flag to package.json, so that it only affects the ESLint repo, but plugins can opt in to enforcing the strict message style if they want.

**Comment by markelog** [comment]
Wow, you right indeed, `RuleTester` is actually public interface. 

> I was thinking maybe add a flag to package.json

Seems a bit radical.

Or we can add additional option to it signature like `strict: true` doesn't seem to be clean solution though since we would need to add it everywhere or make this behaviour by default but wait for major?

**Comment by pmcelhaney** [comment]
Yeah, a flag that would have to be set every time RuleTester is constructed would defeat the purpose. So maybe add a property to the constructor itself?

```js
   RuleTester.isValidMessageFormat = function (message) {
       return message.slice(-1) === ".";
   }
```

**Comment by ilyavolodin** [comment]
I know that keeping all of the error messages in the same style is important, but is it important enough to add something like that to RuleTester, or create additional complexity in some other way?
I think we should just go through and add dots at the end of every message. Sometime in the future we can just do another pass. It's not like it takes all that much time.

**Comment by markelog** [comment]
> but is it important enough to add something like that to RuleTester, or create additional complexity in some other way?

Maybe not important per se, but I think it would be a good thing, from my perspective though not at this point of time, I think in scope of this ticket we can just add periods and probably continue this discussion at 4.0 time, sounds good?

**Comment by ilyavolodin** [comment]
That sounds like a good plan to me.

**Comment by markelog** [comment]
Okay, cool.

@pmcelhaney, @wavebeem would any of you want to help with this one?

**Comment by pmcelhaney** [comment]
Sure, I'll do it this weekend.

**Comment by wavebeem** [comment]
Be my guest; my weekend is looking a little busy. Thanks.

On Fri, Jul 29, 2016 at 2:15 PM, Patrick McElhaney <notifications@github.com
